### PR TITLE
Arbsys timestamp support

### PR DESCRIPTION
**Pull request by hkalodner**
- Updated ArbSys API
  - Replaced `currentMessageTime` with `currentMessageBlock` and `currentMessageTimestamp`
  - Replaced `timeUpperBound` with `blockUpperBound` and `timestampUpperBound`
- Updated ArbSys documentation to add new functions
- Updated Arbitrum time documentation to discuss both block number and timestamp
- Refactored ArbSys in order to remove ugly nesting for switch statement
- Added automatic compile time check that all ArbSys opcodes are implemented
- Replaced hardcoded function identifiers with automatically calculated ones
- Added automated `yarn update-templates` in `contract-templates` which updates `arb-compiler-evm`'s template files

### CODE REVIEW

File: docs/Time_in_Arbitrum.md [hunk 1]
```diff
@@ -4,52 +4,55 @@ title: Time in Arbitrum
...
 
-If your Solidity contract, running on Arbitrum Rollup, accesses `block.number`, this will give you the ArbChain's current time.
-Because ArbChains operate off-chain, an ArbChain's time might differ a little bit from the current Ethereum block time.
-The ArbChain's clock will never get ahead of the Ethereum block time, but the ArbChain's clock might "run a bit slow".
-In particular, it might be as many as 20 blocks behind the true Ethereum time.
+If your Solidity contract, running on Arbitrum Rollup, accesses `block.number` or `block.timestamp`, this will give you the ArbChain's current block or timestamp respectively.
+Because ArbChains operate off-chain, an ArbChain's block number or timestamp might differ a little bit from the current Ethereum block.
+The ArbChain's clock will never get ahead of the Ethereum's, but the ArbChain's clock might "run a bit slow".
+A rollup chain can configure the maximum amount its clock can differ, but generally it will stay within 20 blocks or 10 minutes
+In particular, a rollup chain can configure how far behind it's clock may fall, but generally it will stay within 20 blocks or 10 minutes.
```
Comment (edfelten) [inline docs/Time_in_Arbitrum.md:21]: Typo:  should be "its" instead of "it's".

File: docs/Time_in_Arbitrum.md [hunk 2]
```diff
@@ -4,52 +4,55 @@ title: Time in Arbitrum
...
-At any time, you can get an upper bound on the true Ethereum block time by using this Solidity call:
+At any time, you can get an upper bound on the true Ethereum block number or timestamp by using these Solidity calls:
 
-    ArbSys(address(100)).timeUpperBound()
+    ArbSys(address(100)).blockUpperBound()
+    ArbSys(address(100)).timestampUpperBound()
 
-This will never be more than 20 blocks ahead of `block.number`.
-The real Ethereum block number will always be between the ArbChain's `block.number` and what timeUpperBound() would return (inclusive).
+This will have the same accurancy guarantee from the ArbChain's configuration described about. For example the `blockUpperBound()` can be configured to never be more than 20 blocks ahead of the current Ethereum block number.
```
Comment (edfelten) [inline docs/Time_in_Arbitrum.md:35]: Typo: should be "above" rather than "about".

File: docs/Time_in_Arbitrum.md [hunk 3]
```diff
@@ -4,52 +4,55 @@ title: Time in Arbitrum
...
+Most developers will be fine using `block.number` and `block.timestamp`.
+If you don't mind a little bit of time lag, you can keep on using `block.number` and `block.timestamp` just like you would on Ethereum.
 
 ### Setting deadlines for user response
 
 An exception is a use case like an auction, where the contract emits an event, and you want to give people out there in userland N block times to respond.
 To do that, you'll want to "start the clock ticking" by using code like this:
 
-    deadline = ArbSys(address(100)).timeUpperBound() + N;
+    deadline = ArbSys(address(100)).blockUpperBound() + N;
```
Comment (edfelten) [inline docs/Time_in_Arbitrum.md:51]: Do we want to use timestamps rather than block numbers in this example?

File: docs/Time_in_Arbitrum.md (hunk 2 above)
Comment (hkalodner) [reply docs/Time_in_Arbitrum.md]: Fixed

File: docs/Time_in_Arbitrum.md (hunk 1 above)
Comment (hkalodner) [reply docs/Time_in_Arbitrum.md]: Fixed

File: docs/Time_in_Arbitrum.md (hunk 3 above)
Comment (hkalodner) [reply docs/Time_in_Arbitrum.md]: I think this makes sense. Along with that I changed it to `N hours` rather than `N blocks`

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- packages/arb-compiler-evm/arbitrum/evm/arbsys.py  3
         

Clones added
============
- packages/arb-compiler-evm/arbitrum/tests/test_eth.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5482987&bid=17967621)

**Pull request merged**